crates/bevy_ecs/src/entity/map_entities.rs

```diff
@@ -70,6 +70,55 @@ pub struct EntityMap {
     map: HashMap<Entity, Entity>,
 }
 
+pub struct EntityMapper<'m> {
+    map: &'m mut EntityMap,
+    dead_start: Entity,
+    generations: u32,
+}
+
+impl<'m> EntityMapper<'m> {
+    /// Reserves the allocated dead entities within the world.
+    pub fn save(mut self, world: &mut World) {
+        if self.generations == 0 {
+            return;
+        }
+
+        assert!(world.reserve_generations(self.dead_start, self.generations));
+        self.generations = 0;
+    }
+
+    /// Returns the corresponding mapped entity.
+    pub fn get(&self, entity: Entity) -> Result<Entity, MapEntitiesError> {
+        self.map.get(entity)
+    }
+
+    /// Returns the corresponding mapped entity or allocates a new dead entity if it is absent.
+    pub fn get_or_alloc(&mut self, entity: Entity) -> Entity {
+        if let Ok(mapped) = self.map.get(entity) {
+            return mapped;
+        }
+
+        let new = Entity {
+            generation: self.dead_start.generation + self.generations,
+            index: self.dead_start.index,
+        };
+        self.generations += 1;
+
+        self.map.insert(entity, new);
+
+        new
+    }
+}
+
+impl<'m> Drop for EntityMapper<'m> {
+    fn drop(&mut self) {
+        assert_eq!(
```

Comment:
I'm unsure about panicking here. Can you justify why it needs to panic with a comment?

Reply:
I'll put a comment in the code too, but I was concerned that were `save()` not called and the application to continue, the resulting bugs would be weird and subtle, since you'd end up with entity references pointing at _future_ generations, so in some applications they would initially function alright but suddenly point to a living entity later on in the application. I suspect that would result in quite maddening behavior to root cause, so I wanted to fail-fast and immediately panic rather than leave the application in that state. Does that seem sensible?

Reply:
This is mooted by going with the closure-based API suggested below.